File: examples/python/example_01_window.py
#!/usr/bin/env python

import renderer as rdr

DEFAULT_WIDTH = 1024
DEFAULT_HEIGHT = 768

def main():
    properties = rdr.WindowProperties()
    properties.backend = rdr.WindowBackend.TYPE_GLFW
    properties.width = DEFAULT_WIDTH
    properties.height = DEFAULT_HEIGHT
    properties.title = "Example 01 - Window"

    window = rdr.Window(properties)
    window.RegisterKeyboardCallback(lambda key, action, mods:
        print("key: {}, action: {}. mods: {}".format(key, action, mods)))

    while window.active:
        window.Begin()
        # opengl-stuff can go here :)
        window.End()

if __name__ == '__main__':
    main()

File: examples/cpp/example_01_window.cpp

#include <renderer/engine/graphics/window_t.hpp>
#include <utils/logging.hpp>

constexpr int DEFAULT_WIDTH = 1024;
constexpr int DEFAULT_HEIGHT = 768;

auto main() -> int {
    renderer::WindowConfig win_config;
    win_config.backend = renderer::eWindowBackend::TYPE_GLFW;
    win_config.width = DEFAULT_WIDTH;
    win_config.height = DEFAULT_HEIGHT;
    win_config.title = "Example 01 - Window";

    auto window = ::renderer::Window::Create(win_config);
    window->RegisterKeyboardCallback([&](int key, int action, int mods) {
        LOG_TRACE("key: {0}, action: {1}, mods: {2}", key, action, mods);
        if ((mods & renderer::KEY_MOD_SHIFT) != 0) {
            LOG_INFO("Holding SHIFT key");
        }
        if ((mods & renderer::KEY_MOD_CONTROL) != 0) {
            LOG_INFO("Holding CONTROL key");
        }
        if ((mods & renderer::KEY_MOD_ALT) != 0) {
            LOG_INFO("Holding ALT key");
        }

        if (key == renderer::keys::KEY_ESCAPE) {
            window->RequestClose();
        }
    });

    window->RegisterScrollCallback([](double xOff, double yOff) {
        LOG_TRACE("xOff: {0}, yOff: {1}", xOff, yOff);
    });

    while (window->active()) {
        window->Begin();
        // opengl-stuff can go here :)
        window->End();
    }

    return 0;
}
